samples/matter/common/src/bridge/bridge_manager.cpp

```diff
@@ -108,46 +51,67 @@ CHIP_ERROR BridgeManager::RemoveBridgedDevice(uint16_t endpoint)
 	uint8_t index = 0;
 
 	while (index < kMaxBridgedDevices) {
-		if (nullptr != mBridgedDevices[index]) {
-			if (mBridgedDevices[index]->GetEndpointId() == endpoint) {
+		if (mDevicesMap.Contains(index)) {
+			const DevicePair *devicesPtr = mDevicesMap[index];
+			if ((*devicesPtr).mDevice->GetEndpointId() == endpoint) {
 				LOG_INF("Removed dynamic endpoint %d (index=%d)", endpoint, index);
 				/* Free dynamically allocated memory */
 				emberAfClearDynamicEndpoint(index);
-				Platform::Delete(mDevicesMap[mBridgedDevices[index]]);
-				Platform::Delete(mBridgedDevices[index]);
-				mBridgedDevices[index] = nullptr;
-				return CHIP_NO_ERROR;
+				if (mDevicesMap.Erase(index)) {
+					return CHIP_NO_ERROR;
+				} else {
+					LOG_ERR("Cannot remove bridged devices under index=%d", index);
+					return CHIP_ERROR_NOT_FOUND;
+				}
 			}
 		}
 		index++;
 	}
-
 	return CHIP_ERROR_NOT_FOUND;
 }
 
-CHIP_ERROR BridgeManager::AddDevice(BridgedDevice *device)
+CHIP_ERROR BridgeManager::AddDevices(BridgedDevice *aDevice, BridgedDeviceDataProvider *aDataProvider)
 {
 	uint8_t index = 0;
+	Platform::UniquePtr<BridgedDevice> device(aDevice);
+	Platform::UniquePtr<BridgedDeviceDataProvider> provider(aDataProvider);
+
+	/* Maximum number of Matter bridged devices is controlled inside mDevicesMap,
+	   but the data providers may be created independently, so let's ensure we do not
+	   violate the maximum number of supported instances. */
+	VerifyOrReturnError(++mNumberOfProviders <= kMaxDataProviders, CHIP_ERROR_INTERNAL);
```

Comment:
You mean the if the` mDevicesMap` is full? I agree, we should check that in the first place.

Reply:
I mean that if you increase mNumberOfProviders before checking if it's bigger than kMaxDataProviders, it may grow beyond the maximum (so then even if you release on device, the manager will still think it's full).

Reply:
Fixed.